src/main/scala/io/iohk/ethereum/crypto/zksnark/PairingCheck.scala

```diff
@@ -0,0 +1,169 @@
+package io.iohk.ethereum.crypto.zksnark
+
+import io.iohk.ethereum.crypto.zksnark.BN128.{BN128G1, BN128G2, Point}
+import io.iohk.ethereum.crypto.zksnark.FiniteField.Ops._
+
+import scala.collection.mutable.ArrayBuffer
+
+object PairingCheck {
+
+  val loopCount = BigInt("29793968203157093288")
+
+  /**
+    * Pairing function is defined as: `e: G_1 x G_2 -> G_T` where G1 is element of [[io.iohk.ethereum.crypto.zksnark.BN128.BN128G1]]
+    * and G2 is element of [[io.iohk.ethereum.crypto.zksnark.BN128.BN128G2]]
+    *
+    * Description of algorithms in <a href="https://eprint.iacr.org/2010/354.pdf">optimal ate pairing</a>
+    *
+    * Arithmetic has been ported from <a href="https://github.com/scipr-lab/libff/blob/master/libff/algebra/curves/alt_bn128/alt_bn128_pairing.cpp">libff</a>
+    * Ate pairing algorithms
+    *
+    * @param pairs Sequence of valid pairs of group elements (a1, b1, a2, b2, ..., ak, bk) from (G_1 x G_2)^k^
+    * @return `true` if `log_P1(a1) * log_P2(b1) + ... + log_P1(ak) * log_P2(bk) = 0` else `false`
+    */
+  def pairingCheck(pairs: Seq[G1G2Pair]): Boolean = {
+
+    var product = FiniteField[Fp12].one
+
+    pairs.foreach{ pair =>
```

Comment:
Maybe it will be nicer: `pairs.map(p=> millerLoop(p.a, p.b).filter(_ != FiniteField.one).fold(FiniteField.one)(_ *_)`

Reply:
I do not like how your approach creates two unnecessary collections (after map and filter), but definitely i can improve my approach by using fold instead of `var` and `foreach`

Reply:
Maybe you could use views or sth like that